packages/netstring/writer.js

```diff
@@ -39,14 +41,28 @@ export const makeNetstringWriter = (output, { chunked = false } = {}) => {
       const prefix = getLengthPrefixCharCodes(messageLength);
 
       if (chunked) {
-        return Promise.all([
+        const partsWritten = [
           output.next(new Uint8Array(prefix)),
-          ...messageChunks.map(async chunk => output.next(chunk)),
+          ...messageChunks.map(chunk => output.next(chunk)),
           output.next(COMMA_BUFFER),
-        ]).then(([r1, r2, r3]) => ({
-          done: !!(r1.done || r2.done || r3.done),
-          value: undefined,
-        }));
+        ];
+
+        const writerClosedEarly = makePromiseKit();
+        for (const promise of partsWritten) {
+          promise.then(result => {
+            if (result.done) {
+              writerClosedEarly.resolve(result);
+            }
+          });
+        }
+
+        return Promise.race([
+          writerClosedEarly.promise,
+          Promise.all(partsWritten).then(results => ({
+            done: results.some(({ done }) => done),
+            value: undefined,
+          })),
```

Comment:
It feels like we're racing ourselves here, and the whole "check done again" is to make sure we resolve with `false` if one of the stream has closed?